pulsar-broker/src/main/java/com/yahoo/pulsar/broker/loadbalance/impl/ModularLoadManagerImpl.java

```diff
@@ -357,16 +357,24 @@ private void updateAllBrokerData() {
                         // Replace previous local broker data.
                         brokerDataMap.get(broker).setLocalData(localData);
                     } else {
-                        // Initialize BrokerData object for previously unseen
-                        // brokers.
+                        // Initialize BrokerData object for previously unseen brokers.
                         brokerDataMap.put(broker, new BrokerData(localData));
                     }
                 } catch (Exception e) {
                     log.warn("Error reading broker data from cache for broker - [{}], [{}]", broker, e.getMessage());
                 }
             }
+            // Remove obsolete brokers.
+            for (final String broker : brokerDataMap.keySet()) {
```

Comment:
I think it could even still be an issue if the broker becomes unavailable while filtering policies and selecting candidates, though it would be impractical to handle this case perfectly so I will reduce synchronization and use `availableActiveBrokers` to give the best likelihood of selecting an active broker.